src/adapter/src/coord/statement_logging.rs

```diff
@@ -381,33 +385,52 @@ impl Coordinator {
         vec![(retraction, -1), (new, 1)]
     }
 
-    /// Set the `cluster_id` for a statement, once it's known.
-    pub fn set_statement_execution_cluster(
+    /// Mutate a statement execution record via the given function `f`.
+    fn mutate_record<F: FnOnce(&mut StatementBeganExecutionRecord)>(
         &mut self,
         StatementLoggingId(id): StatementLoggingId,
-        cluster_id: ClusterId,
+        f: F,
     ) {
-        let cluster_name = self.catalog().get_cluster(cluster_id).name.clone();
         let record = self
             .statement_logging
             .executions_begun
             .get_mut(&id)
-            .expect("set_statement_execution_cluster must not be called after execution ends");
-        if record.cluster_id == Some(cluster_id) {
-            return;
-        }
+            .expect("mutate_record must not be called after execution ends");
         let retraction = Self::pack_statement_began_execution_update(record);
         self.statement_logging
             .pending_statement_execution_events
             .push((retraction, -1));
-        record.cluster_name = Some(cluster_name);
-        record.cluster_id = Some(cluster_id);
+        f(record);
         let update = Self::pack_statement_began_execution_update(record);
         self.statement_logging
             .pending_statement_execution_events
             .push((update, 1));
     }
 
+    /// Set the `cluster_id` for a statement, once it's known.
+    pub fn set_statement_execution_cluster(
+        &mut self,
+        id: StatementLoggingId,
+        cluster_id: ClusterId,
+    ) {
+        let cluster_name = self.catalog().get_cluster(cluster_id).name.clone();
+        self.mutate_record(id, |record| {
+            record.cluster_name = Some(cluster_name);
+            record.cluster_id = Some(cluster_id);
+        });
+    }
+
+    /// Set the `execution_timestamp` for a statement, once it's known
+    pub fn set_statement_execution_timestamp(
+        &mut self,
+        id: StatementLoggingId,
+        timestamp: Timestamp,
+    ) {
+        self.mutate_record(id, |record| {
```

Comment:
do we expect the cost of extra retractions and consolidate to be not a problem here?

Reply:
I don't think it will matter, I think the situations where this is called twice for the same statement are kinda unusual anyway.